paasta_tools/check_autoscaler_max_instances.py

```diff
@@ -0,0 +1,159 @@
+from paasta_tools.utils import DEFAULT_SOA_DIR
+from paasta_tools.utils import list_services
+from paasta_tools.utils import load_system_paasta_config
+from paasta_tools.utils import SystemPaastaConfig
+
+log = logging.getLogger(__name__)
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description=(
+            "Check all autoscaled services to see if they're at their max_instances."
+            " If so, send an alert if their utilization is above max_instances_alert_threshold."
+        )
+    )
+    parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        default=DEFAULT_SOA_DIR,
+        help="Use a different soa config directory",
+    )
+    parser.add_argument(
+        "--dry-run",
+        dest="dry_run",
+        action="store_true",
+        help="Print Sensu alert events instead of sending them",
+    )
+    return parser.parse_args()
+
+
+async def check_max_instances(
+    soa_dir: str,
+    cluster: str,
+    instance_type_class: Type[KubernetesDeploymentConfig],
+    system_paasta_config: SystemPaastaConfig,
+    dry_run: bool = False,
+):
+    kube_client = KubeClient()
+    for service in list_services(soa_dir=soa_dir):
+        service_config = PaastaServiceConfigLoader(service=service, soa_dir=soa_dir)
+        for job_config in service_config.instance_configs(
+            cluster=cluster, instance_type_class=instance_type_class
+        ):
+            instance = job_config.get_instance()
+            if not job_config.get_autoscaling_metric_spec(
+                name=get_kubernetes_app_name(service, instance),
+                cluster=cluster,
+                kube_client=kube_client,
+                namespace=job_config.get_namespace(),
+            ):
+                # Not an instance that uses HPA, don't check.
+                # TODO: should we send status=0 here, in case someone disables autoscaling for their service / changes
+                # to bespoke autoscaler?
+                continue
+
+            if not job_config.get_docker_image():
+                # skip services that haven't been marked for deployment yet.
+                continue
+
+            autoscaling_status = await pik.autoscaling_status(
+                kube_client=kube_client,
+                job_config=job_config,
+                namespace=job_config.get_namespace(),
+            )
+            if autoscaling_status["min_instances"] == -1:
+                log.warning(
+                    f"HPA {job_config.get_sanitised_deployment_name()} not found."
+                )
+                continue
+
+            if (
+                autoscaling_status["min_instances"]
+                == autoscaling_status["max_instances"]
+            ):
+                status = 0
+                output = f"Not checking {service}.{instance} as min_instances == max_instances == {autoscaling_status['max_instances']}. This is probably a canary instance."
```

Comment:
ah, would this be when folks use yaml anchors and have the canary reuse the main config?

Reply:
(i would not expect any canaries to have an hpa)

Reply:
i do wonder if maybe there could be a case where someone overrode max_instances and autotune has raised min_instances to the current max_instance override

Reply:
> ah, would this be when folks use yaml anchors and have the canary reuse the main config?

Yeah, yelp-main's canaries use this pattern.

> i do wonder if maybe there could be a case where someone overrode max_instances and autotune has raised min_instances to the current max_instance override

Hmm, good point. There are a couple options:

- look for `canary` in the instance name
- look for other instances in the same cluster with the same `registrations`
- check to make sure min_/max_instances are both manually set, not set in autotune_defaults

Reply:
yea, not sure what I'd pick if I was making this change - looking for `canary` in the name is maybe the option least likely to inadvertently be triggered?